### PR TITLE
[SPARK-46069][SQL] Support unwrap timestamp type to date type

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Just like [[SPARK-42597][SQL] Support unwrap date type to timestamp type](https://github.com/apache/spark/pull/40190), this PR enhance `UnwrapCastInBinaryComparison` to support unwrap timestamp type to date type.

Add two new expressions:
1. floorDate: the largest date that is less than or equal to the input timestamp.
2. dateAddOne: floorDate + one day.
3. isStartOfDay: return true if ts == floorDate

The way to unwrap timestamp type to date type are:

1. CAST(date AS timestamp) > ts ===> date > floorDate
2. CAST(date AS timestamp) >= ts ===> if(isStartOfDay) date >= floorDate else date >= dateAddOne
3. CAST(date AS timestamp) === ts ===>
        if (isStartOfDay) {
          fromExp === floorDate
        } else if (!fromExp.nullable) {
          FalseLiteral
        } else {
          fromExp === floorDate AND fromExp === dateAddOne
        }
4. CAST(date AS timestamp) <=> ts ===> if (isStartOfDay) date <=> floorDate else FalseLiteral
5. CAST(date AS timestamp) < ts ===> if(isStartOfDay) date < floorDate else date < dateAddOne
6. CAST(date AS timestamp) <= ts ===> date <= floorDate

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @sunchao FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 1]
```diff
@@ -329,6 +334,40 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
...
+    val floorDate = Cast(ts, fromExp.dataType, tz, evalMode)
+    val equalToFloorDateExpr =
+      EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode))
+    val equalToFloorDate = equalToFloorDateExpr.eval(EmptyRow).asInstanceOf[Boolean]
+    val ceilDate =
+      if (equalToFloorDate) {
+        floorDate
+      } else {
+        DateAdd(floorDate, Literal(1, IntegerType))
+      }
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:352]:
```suggestion
    val floorDate = Cast(ts, fromExp.dataType, tz, evalMode)
    val isStartOfDay =
      EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode)).eval(EmptyRow).asInstanceOf[Boolean]
    val ceilDate = if (isStartOfDay) floorDate else DateAdd(floorDate, Literal(1, IntegerType))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 2]
```diff
@@ -329,6 +334,40 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
...
+
+    exp match {
+      case _: GreaterThan =>
+        GreaterThan(fromExp, floorDate)
+      case _: GreaterThanOrEqual =>
+        GreaterThanOrEqual(fromExp, ceilDate)
+      case _: EqualTo =>
+        And(EqualTo(fromExp, floorDate), EqualTo(fromExp, ceilDate))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:360]:
```suggestion
        if (isStartOfDay) EqualTo(fromExp, floorDate) else FalseLiteral
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 3]
```diff
@@ -329,6 +334,40 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
...
+      case EqualNullSafe(_, _) =>
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:361]: It seems the left side should not be nullable.
```suggestion
      case EqualNullSafe(left, _) if !left.nullable =>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala (hunk 2 above)
Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:354]: Thanks @wangyum for your review.

If `isStartOfDay` is false, the optimizer will change this expression to `FalseLiteral` which results in incorrect result if `fromExp` is null.
So I use two EqualTo expressions to support nullable input expressions. During the execution phase, if `isStartOfDay` is true, the two EqualTo expressions can be reuse and no additional calculations are added.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:369]: How about?
```scala
        if (isStartOfDay) {
          EqualTo(fromExp, floorDate)
        } else if (!fromExp.nullable) {
          FalseLiteral
        } else {
          And(EqualTo(fromExp, floorDate), EqualTo(fromExp, ceilDate))
        }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:370]: How about?
```scala
        if (isStartOfDay) EqualNullSafe(fromExp, floorDate) else FalseLiteral
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala (hunk 2 above)
Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala]: Updated, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala (hunk 3 above)
Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:376]: Updated, thanks

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
cc @cloud-fan

**Comment by wangyum** [comment]
Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 4]
```diff
@@ -138,6 +138,11 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if AnyTimestampType.acceptsType(fromExp.dataType) && value != null =>
       Some(unwrapDateToTimestamp(be, fromExp, date, timeZoneId, evalMode))
 
+    case be @ BinaryComparison(
+      Cast(fromExp, _, timeZoneId, evalMode), ts @ Literal(value, _))
```
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:142]: e.g., `fromExp: DateType`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 5]
```diff
@@ -329,6 +334,48 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     }
   }
 
+  private def unwrapTimeStampToDate(
```
[the remark this answers is not in the document]
Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:337]: Sorry, fix it in the followup PR : https://github.com/apache/spark/pull/44134/files#diff-56271650e1a2efb907de03cfd8300018b021dd7ae338f13c6d6e6448cd2bac8aL144

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 6]
```diff
@@ -329,6 +334,48 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
...
+        } else {
+          GreaterThanOrEqual(fromExp, dateAddOne)
+        }
+      case _: EqualTo =>
+        if (isStartOfDay) {
+          EqualTo(fromExp, floorDate)
+        } else if (!fromExp.nullable) {
+          FalseLiteral
+        } else {
+          And(EqualTo(fromExp, floorDate), EqualTo(fromExp, dateAddOne))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:363]: This is very confusing. How can the `fromExp` be equal to two different values at the same time? This expression is either false or null.

Do you mean `if(isnull(e), null, false)`? Then `And(IsNull(e), Literal(null, BooleanType))` should be more efficient.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:363]: actually there is already a `falseIfNotNull` method in this class.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:363]: BTW let's keep the high code review quality in the community. This is the core logic of this optimization and it's even highlighted in the PR description. I'm surprised we missed this case.

**Comment by gatorsmile** [comment]
@wangyum  Please be more careful in the code review.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala [hunk 7]
```diff
@@ -403,11 +405,40 @@ class UnwrapCastInBinaryComparisonSuite extends PlanTest with ExpressionEvalHelp
       castDate(f5) > nullLit || castDate(f6) > nullLit)
   }
 
+  test("SPARK-46069: Support unwrap timestamp type to date type") {
+    val tsLit = Literal.create(ts1, TimestampType)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala:409]: > 3. isStartOfDay: return true if ts == floorDate

For such an important concept, we don't have unit tests for it? The optimized comparison is different depending on this condition!

**Comment by wankunde** [comment]
@wangyum @cloud-fan 

I'm sorry, it seems  the expression `val isStartOfDay = EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode)).eval(EmptyRow).asInstanceOf[Boolean]`  is incorrect.
For example, the input ts is `2023-01-01 00:00:00`,  floorDate will be `2022-12-31 16:00:00`, and `Cast(floorDate, ts.dataType, tz, evalMode)` will be `2022-12-31 08:00:00`, I don't know why...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 8]
```diff
@@ -329,6 +334,48 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
...
+  private def unwrapTimeStampToDate(
+      exp: BinaryComparison,
+      fromExp: Expression,
+      ts: Literal,
+      tz: Option[String],
+      evalMode: EvalMode.Value): Expression = {
+    val floorDate = Cast(ts, fromExp.dataType, tz, evalMode)
+    val dateAddOne = DateAdd(floorDate, Literal(1, IntegerType))
+    val isStartOfDay =
+      EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode)).eval(EmptyRow).asInstanceOf[Boolean]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:346]: @wankunde  do you mean the cast here does not produce a timestamp with the time part to be 0 (00:00:00)?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:346]: how did you print it? Do you use Spark Cast (cast to string) to print it?

Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:346]: Add a followup for this PR:  https://github.com/apache/spark/pull/44366

**Comment by wangyum** [comment]
> @wangyum Please be more careful in the code review.

Sorry. Will be careful in future code review.

**Pull request closed without merging**